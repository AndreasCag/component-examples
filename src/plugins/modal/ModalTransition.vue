<template>
  <transition
    class="modal-transition"
    name="modal"
    appear
    v-on="$listeners"
  >
    <slot />
  </transition>
</template>

<style scoped>
.modal-enter,
.modal-leave-to {
  opacity: 0;
}

.modal-enter-active,
.modal-leave-active {
  transition: opacity .3s ease;
}
</style>
