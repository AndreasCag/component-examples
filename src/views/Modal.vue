<template>
  <div>
    <div>
      <button @click="canShowFirstModal = true;">
        Show modal
      </button>
    </div>

    <BeautyModal
      v-if="canShowFirstModal"
      :actions="firstModalActions"
      @close="canShowFirstModal = false;"
    >
      <template #header>
        Lorem ipsum
      </template>
      <template #content>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit?
      </template>
    </BeautyModal>

    <BeautyModal
      v-if="canShowSecondModal"
      :actions="secondModalActions"
      @close="canShowSecondModal = false;"
    >
      <template #content>
        Lorem ipsum dolor sit amet!
      </template>
    </BeautyModal>
  </div>
</template>

<script>
import BeautyModal from '@/components/BeautyModal/index.vue';

export default {

  components: {
    BeautyModal,
  },

  data: () => ({
    canShowFirstModal: false,
    canShowSecondModal: false,
  }),

  computed: {

    firstModalActions() {
      return [
        {
          text: 'no',
          action: () => {
            // eslint-disable-next-line vue/no-side-effects-in-computed-properties
            this.canShowFirstModal = false;
          },
        },
        {
          text: 'yes',
          mode: 'primary',
          action: () => {
            // eslint-disable-next-line vue/no-side-effects-in-computed-properties
            this.canShowSecondModal = true;
          },
        },
      ];
    },

    secondModalActions() {
      return [
        {
          text: 'close',
          mode: 'primary',
          action: () => {
            // eslint-disable-next-line vue/no-side-effects-in-computed-properties
            this.canShowSecondModal = false;
          },
        },
      ];
    },

  },

};
</script>
