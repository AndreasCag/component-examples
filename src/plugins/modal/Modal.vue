<script>
import ModalTransition from './ModalTransition.vue';

export default {

  props: {
    renderContent: {
      type: Function,
      required: true,
    },
  },

  data: () => ({
    // isStartClosing is used to play close animation
    isStartClosing: false,
  }),

  methods: {

    backgroundIsClicked(event) {
      if (event.target !== event.currentTarget) {
        return;
      }

      this.$emit('close');
    },

    closeModal() {
      this.isStartClosing = true;
    },

    emitLeaveTransitionEnd() {
      this.$emit('leaveTransitionEnd');
    },

  },

  render() {
    const { isStartClosing } = this;

    return (
      <ModalTransition
        onAfterLeave={this.emitLeaveTransitionEnd}
      >
        {
          !isStartClosing ? (
            <div
              class="background"
              onClick={this.backgroundIsClicked}
            >
              <div class="container">
                {this.renderContent()}
              </div>
            </div>
          ) : null
        }
      </ModalTransition>
    );
  },

};
</script>

<style scoped>
.background {
  position: fixed;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, .3);
}

.container {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  background-color: white;
  border-radius: 2px;
}
</style>
