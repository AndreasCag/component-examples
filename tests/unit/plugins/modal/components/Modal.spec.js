import Modal from '@/plugins/modal/Modal.vue';
import ModalTransition from '@/plugins/modal/ModalTransition.vue';
import { shallowMount } from '@vue/test-utils';
import Vue from 'vue';

// H variable is used in JSX
// eslint-disable-next-line no-unused-vars
const h = new Vue().$createElement;

describe('Modal.vue', () => {
  it('renders pasted content', () => {
    const modalWrapper = shallowMount(Modal, {
      propsData: {
        renderContent: () => [
          <div class="test-content">
            Test content
          </div>,
          <div class="another-test-content">
            Another test content
          </div>,
        ],
      },
    });

    expect(modalWrapper.element).toMatchSnapshot();
  });

  it('emits close event after click event on background', () => {
    const modalWrapper = shallowMount(Modal, {
      propsData: {
        renderContent: () => null,
      },
    });

    modalWrapper.find('.background').trigger('click');

    expect(modalWrapper.emitted().close.length).toBe(1);
  });

  it('emits leaveTransitionEnd after afterLeave event in ModalTransition', async () => {
    const modalWrapper = shallowMount(Modal, {
      propsData: {
        renderContent: () => null,
      },
    });

    modalWrapper.find(ModalTransition).vm.$emit('afterLeave');

    expect(modalWrapper.emitted().leaveTransitionEnd.length).toBe(1);
  });
});
