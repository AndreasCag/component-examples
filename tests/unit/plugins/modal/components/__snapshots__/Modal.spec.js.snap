// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal.vue renders pasted content 1`] = `
<anonymous-stub>
  <div
    class="background"
  >
    <div
      class="container"
    >
      <div
        class="test-content"
      >
        Test content
      </div>
      <div
        class="another-test-content"
      >
        Another test content
      </div>
    </div>
  </div>
</anonymous-stub>
`;
